llvm/lib/Transforms/Yk/ShadowStack.cpp

```diff
@@ -271,6 +275,74 @@ class YkShadowStack : public ModulePass {
     }
   }
 
+  // Handle `main` separatley, since it works slightly differently to other
+  // functions: it allocates the shadow stack
+  //
+  // For `__yk_unopt_main` we update the allocas while reusing the same shadow
+  // stack instance allocated by `main`.
+  //
+  // Note that since we assuming main() doesn't call main(), we can consider
+  // the shadow stack disused at the point main() returns. For this reason,
+  // there's no need to emit a shadow epilogue for main().
+  //
+  // YKFIXME: Investigate languages that don't have/use main as the first
+  // entry point.
+  void updateMainFunctions(DataLayout &DL, Module &M, GlobalVariable *SSGlobal,
+                           LLVMContext &Context) {
+    Function *Main = M.getFunction(MAIN);
+    if (Main == nullptr || Main->isDeclaration()) {
+      Context.emitError("Unable to add shadow stack: could not find definition "
+                        "of \"main\" function!");
+      return;
+    }
+
+    // Update the original main function.
+    AllocaVector MainAllocas;
+    std::vector<ReturnInst *> MainRets;
+    size_t SFrameSize = analyseFunction(*Main, DL, MainAllocas, MainRets);
+    CallInst *Malloc = insertMainPrologue(Main, SSGlobal, SFrameSize);
+    rewriteAllocas(DL, MainAllocas, Malloc);
+
+    // If we have two control points, we need to update the cloned main
+    // function as well.
+    if (controlPointCount == 2) {
+      Function *UnoptMain = M.getFunction(YK_UNOPT_MAIN);
+      if (UnoptMain == nullptr || UnoptMain->isDeclaration()) {
+        Context.emitError(
+            "Unable to add shadow stack: could not find definition "
+            "of \"__yk_unopt_main\" function!");
+        return;
+      }
+
+      AllocaVector UnoptMainAllocas;
+      std::vector<ReturnInst *> UnoptMainRets;
+      size_t SFrameSizeUnopt =
+          analyseFunction(*UnoptMain, DL, UnoptMainAllocas, UnoptMainRets);
+
+      // Insert a load of SSGlobal at the beginning of UnoptMain.
+      BasicBlock &EntryBB = UnoptMain->getEntryBlock();
+      Instruction *FirstNonPHI = EntryBB.getFirstNonPHI();
+      IRBuilder<> Builder(FirstNonPHI);
+
+      // Load the current shadow stack pointer from the global variable.
+      LoadInst *LoadedSSPtr = Builder.CreateLoad(Int8PtrTy, SSGlobal);
+
+      // Assert that the allocas count match between opt/unopt main
+      assert(MainAllocas.size() == UnoptMainAllocas.size() &&
+             "Expected same number of allocas between opt/unopt main");
+
+      // Assert that the allocas have the same offset
+      for (size_t i = 0; i < MainAllocas.size(); i++) {
+        auto &[optAI, optOffset] = MainAllocas[i];
+        auto &[unoptAI, unoptOffset] = UnoptMainAllocas[i];
+        assert(optOffset == unoptOffset && "Alloca offsets must match");
```

Comment:
Any other values that need to be asserted here?
I noticed that the types are not always the same...
This assertion fails:
```cpp
assert(optAI->getAllocatedType() == unoptAI->getAllocatedType() &&
               "Alloca types must match");
```

Reply:
I think that should do. But it's weird that the types don't match. I wasn't expecting that to happen.

Reply:
ok, yeah it's only in on specific yk test `ang_tests::sdiv.c`
```
Mismatched alloca types at index 7:
  Optimized alloca type: 0x70
  Unoptimized alloca type: 0x55e49c008e90
```

Reply:
It was a weird one, added fix and explanation comment here :point_right: 6aff8e933d3847f940055bf5944719c0568c329a
But then I realised that we should assert on the replaced instructions instead of the original allocas :point_right: 2c492ed7e01bea7250dbec9fd0b7c3cb146e346c